### DISCUSSION TITLE
3D grid shown out of actual working space.

### DISCUSSION THREAD

**Discussion by 300bar** [terjeio/ioSender#306] [General]
Hi Terje, what a nice g-code sender!
My project (pcb milling machine) is finish and working. I would only like to iron a detail, and it is the working grid visibility, maybe I am missing something but can´t find it. When homing the machine (top-right) to zero Mpos, and defining the xyz max travel in settings (it only accepts positive numbers) the 3D grid is displayed as positive machine coordinates but obviously I am working in negative machines coordinates. Everything works Ok hard limits, soft limits, homing, work coordinates… everything except the 3d grid (and machine work envelopment) that is shown in 3D out the real working space. See picture. 
I would appreciate any input.

![Captura de pantalla 2023-06-26 130841](https://github.com/terjeio/ioSender/assets/64420106/50a702f9-aa1b-4677-b851-87004888ae22)

**Comment by terjeio** [comment]
If you display the machine work envelope (click the larger blue rectangle) is the grid inside of that?
G54 offset is 0,0,0?

The grid displays correctly for me, and 0,0 is in the top right corner:

![image](https://github.com/terjeio/ioSender/assets/20260062/d78c6444-be1f-40d4-871d-6761a28cdbab)

**Comment by 300bar** [comment]
Hi Terje,

 

Attached see a screen, yes the working space appears like yours, it is only my working space out of it.

 

Thank you for the help.

 

Luis.

Madrid.

 

 

De: Terje Io ***@***.***> 
Enviado el: martes, 27 de junio de 2023 12:06
Para: terjeio/ioSender ***@***.***>
CC: 300bar ***@***.***>; Author ***@***.***>
Asunto: Re: [terjeio/ioSender] 3D grid shown out of actual working space. (Discussion #306)

 

If you display the machine work envelope (click the larger blue rectangle) is the grid inside of that?
G54 offset is 0,0,0?

The grid displays correctly for me, and 0,0 is in the top right corner:

 <https://user-images.githubusercontent.com/20260062/249115198-d78c6444-be1f-40d4-871d-6761a28cdbab.png>

**Comment by 300bar** [comment]
![Captura de pantalla 2023-06-27 121753](https://github.com/terjeio/ioSender/assets/64420106/6dade7d8-dceb-47ca-be7d-1fcbadaab53c)

**Reply by terjeio** [reply]
Do you have a G54 offset active? Check in the _Offsets_ tab.

**Comment by 300bar** [comment]
Yes I have x,y in negative

 

 

De: Terje Io ***@***.***> 
Enviado el: martes, 27 de junio de 2023 13:05
Para: terjeio/ioSender ***@***.***>
CC: 300bar ***@***.***>; Comment ***@***.***>
Asunto: Re: [terjeio/ioSender] 3D grid shown out of actual working space. (Discussion #306)

 

Do you have a G54 offset active? Check in the Offsets tab.

**Comment by terjeio** [comment]
If you clear the G54 offset and move to X and Y negative max travel and then zero X and Y do you get the job inside the machine work envelope?

**Reply by 300bar** [reply]
I have cleared G54 offsets after homing, and the cone is in 0 machine and 0 work coordinates inside the grid. I load a gcode and it appears inside the grid as well, then I move the spindle to the real 0 work coordinates  **_before_** setting the work coodinates to 0 and I get the following screen
![Captura de pantalla 2023-06-28 091918](https://github.com/terjeio/ioSender/assets/64420106/214f4b4c-e475-413c-a564-47d545d86be1)

**Reply by terjeio** [reply]
> then I move the spindle to the real 0 work coordinates before setting the work coodinates to 0

The move is with G53? And do you have a G92 offset active as well?

Hard to tell what is going on without knowing all the details such as controller settings...

**Comment by 300bar** [comment]
Hi Terje,
No G53 nor G92…
I have a theory, my machine when starting up first thing to do is homing, once homed, the grid does not display although the working space in reference to the cone is settled and it couls be shown. Ans here is the strange thing, I load a gcode, then the 3D representation appears, BUT the cone is not in the correct corner, as mi homing is done in the top right end, the cone, with no offsets should be represented in the same corner as the homing but is it shown in the bottom left.  
So could happen than due to my settings, inverting some direction signals to match the wiring of my steppers, affects the 3D representation?????

**Reply by terjeio** [reply]
That is why I wanted your controller settings. _Help > About >To clipboard_ and paste in a comment.

**Comment by 300bar** [comment]
%
; grblHAL
; 1.1f.20230526
; [OPT:VNMZ0SL+2,35,1024,3,0]
; [NEWOPT:ENUMS,RT+,HOME,ES,REBOOT,TC,SED,RTC]
; [FIRMWARE:grblHAL]
; [NVS STORAGE:*FLASH]
; [DRIVER:iMXRT1062]
; [DRIVER VERSION:230526]
; [DRIVER OPTIONS:USB.1]
; [BOARD:T41U5XBB]
; [AUX IO:4,3,0,0]
; [PLUGIN:KEYPAD v1.34]
;
$N0=
$N1=
; 0 - Step pulse time
$0=5.0
; 1 - Step idle delay
$1=25
; 2 - Step pulse invert
$2=7
; 3 - Step direction invert
$3=2
; 4 - Invert stepper enable pin(s)
$4=0
; 5 - Invert limit pins
$5=3
; 6 - Invert probe pin
$6=1
; 8 - Ganged axes direction invert
$8=0
; 9 - PWM Spindle
$9=3
; 10 - Status report options
$10=511
; 11 - Junction deviation
$11=0.010
; 12 - Arc tolerance
$12=0.002
; 13 - Report in inches
$13=0
; 14 - Invert control pins
$14=78
; 15 - Invert coolant pins
$15=0
; 16 - Invert spindle signals
$16=0
; 17 - Pullup disable control pins
$17=0
; 18 - Pullup disable limit pins
$18=0
; 19 - Pullup disable probe pin
$19=0
; 20 - Soft limits enable
$20=1
; 21 - Hard limits enable
$21=1
; 22 - Homing cycle
$22=41
; 23 - Homing direction invert
$23=0
; 24 - Homing locate feed rate
$24=100.0
; 25 - Homing search seek rate
$25=2000.0
; 26 - Homing switch debounce delay
$26=50
; 27 - Homing switch pull-off distance
$27=2.000
; 28 - G73 Retract distance
$28=1.000
; 29 - Pulse delay
$29=5.0
; 30 - Maximum spindle speed
$30=10000.000
; 31 - Minimum spindle speed
$31=0.000
; 32 - Mode of operation
$32=0
; 33 - Spindle PWM frequency
$33=50
; 34 - Spindle PWM off value
$34=5.0
; 35 - Spindle PWM min value
$35=5.0
; 36 - Spindle PWM max value
$36=7.5
; 37 - Steppers deenergize
$37=7
; 39 - Enable legacy RT commands
$39=1
; 40 - Limit jog commands
$40=1
; 41 - Parking cycle
$41=0
; 42 - Parking axis
$42=2
; 43 - Homing passes
$43=1
; 44 - Axes homing, first pass
$44=4
; 45 - Axes homing, second pass
$45=3
; 46 - Axes homing, third pass
$46=0
; 50 - Step jog speed
$50=500.0
; 51 - Slow jog speed
$51=500.0
; 52 - Fast jog speed
$52=2000.0
; 53 - Step jog distance
$53=0.100
; 54 - Slow jog distance
$54=500.0
; 55 - Fast jog distance
$55=300.0
; 56 - Parking pull-out distance
$56=5.0
; 57 - Parking pull-out rate
$57=100.0
; 58 - Parking target
$58=-5.0
; 59 - Parking fast rate
$59=500.0
; 60 - Restore overrides
$60=0
; 61 - Safety door options
$61=0
; 62 - Sleep enable
$62=0
; 63 - Feed hold actions
$63=3
; 64 - Force init alarm
$64=0
; 65 - Probing feed override
$65=0
; 100 - X-axis travel resolution
$100=800.000
; 101 - Y-axis travel resolution
$101=800.000
; 102 - Z-axis travel resolution
$102=800.000
; 110 - X-axis maximum rate
$110=2000.000
; 111 - Y-axis maximum rate
$111=2000.000
; 112 - Z-axis maximum rate
$112=500.000
; 120 - X-axis acceleration
$120=200.000
; 121 - Y-axis acceleration
$121=200.000
; 122 - Z-axis acceleration
$122=200.000
; 130 - X-axis maximum travel
$130=102.000
; 131 - Y-axis maximum travel
$131=240.000
; 132 - Z-axis maximum travel
$132=18.000
; 341 - Tool change mode
$341=0
; 342 - Tool change probing distance
$342=30.0
; 343 - Tool change locate feed rate
$343=25.0
; 344 - Tool change search seek rate
$344=200.0
; 345 - Tool change probe pull-off rate
$345=200.0
; 346 - Restore position after M6
$346=1
; 370 - Invert I/O Port inputs
$370=0
; 372 - Invert I/O Port outputs
$372=0
; 384 - Disable G92 persistence
$384=0
; 392 - Spindle on delay
$392=4.0
; 393 - Coolant on delay
$393=1.0
; 398 - Planner buffer blocks
$398=35
; 481 - Autoreport interval
$481=0
%

**Comment by terjeio** [comment]
> ... the 3D grid is displayed as positive machine coordinates but obviously I am working in negative machines coordinates. 

![image](https://github.com/terjeio/ioSender/assets/20260062/c68660c5-7d75-4e8d-b395-5ccd696f5028)

You have configured the controller for positive coordinates, uncheck _Set machine origin to 0_ and rehome for negative. Restart the sender afterwards so that it reconfigures itself accordingly.

**Comment by 300bar** [comment]
Ok, unchecking "set machine to 0" alone works the trick and the grid displays correctly, but now al G53 (move in machine coordinates) does not work and the macros using G53 neither. Machine coordinates must be set to 0 at homing, in my opinion.

**Reply by terjeio** [reply]
> G53 (move in machine coordinates) does not work and the macros using G53 neither.

"does not work" is not very helpful if there are bugs. You get error messages? Examples?

> Machine coordinates must be set to 0 at homing, in my opinion.

Negative or positive machine space should both work.

**Comment by 300bar** [comment]
Sorry Terje, you are right I was not very clear in my last message.  What happens is the following.:
 Uncheking "Set machine origin to 0" makes the grid appears correctly. Homing works well but as I have 1mm pull off distance from the limit switch once the homeming stops I get Mpos (-1,-1,-1) and the grid if correct. The problem for example is if I perform a G28 (Goto Mpos(0,0,0)) This shows up as out of the machine limits and triggers an alarm, while when cheking "Set machine origin to 0" after homing the Mpos was (0,0,0) and G28 worked. When I said before G53 was not working was because some of my macros were failing se same way as G28.

I hope this helps. 





2

**Reply by terjeio** [reply]
[Program the G28 offset](http://linuxcnc.org/docs/devel/html/gcode/g-code.html#gcode:g28-g28.1) to be off the limit switches.
It is unfortunate that legacy Grbl set G28 (and G30) to 0,0,0 - the default values should, IMO, be undefined and G28 should move to the machine origin as specified. It is not trivial to change this as it is likely it will mess up senders.

**Comment by 300bar** [comment]
Thank you very much for the explanation.
 If I check, as I had, the "Set machine origin to 0", then once homed Mpos will be (0,0,0) and everything will perform well, is it possible for the sender to take this checked flag into consideration and draw the grid accordingly?

**Reply by terjeio** [reply]
> is it possible for the sender to take this checked flag into consideration and draw the grid accordingly?

The flag is taken into consideration already - not sure what you want to be handled differently.

FYI I have uploaded a [new edge version](http://www.io-engineering.com/downloads) with an additional grid mode, displayed when the machine is not homed. In addition I have disabled the show work envelope button when this mode is active as showing it is not possible/meaningful unless the machine is homed.
